[Cloud Deployment IIIb]: AWS batch deployment helper

CodyCBakerPhD opened the pull request:
Example prototype for easy deployment of a docker image on AWS Batch

CodyCBakerPhD commented:
(various changes particular to your compute environments will probably be necessary)

This also uses EFS instead of EBS, which is more expensive per allocation but allows completely dynamic sizing in that way

bendichter commented on src/neuroconv/tools/data_transfers.py:
```diff
@@ -349,3 +351,154 @@ def automatic_dandi_upload(
+
+def submit_aws_batch_job(
+    job_name: str,
+    docker_container: str,
```
```suggestion
    docker_container: str = ghcr.io/catalystneuro/rclone_auto_config:latest,
```

CodyCBakerPhD commented on tests/test_minimal/test_tools/s3_tools.py:
```diff
@@ -46,3 +47,37 @@ def test_estimate_total_conversion_runtime():
+    docker_image = "ubuntu:latest"
+    command = "echo 'Testing NeuroConv AWS Batch submission."
+
+    submit_aws_batch_job(
+        job_name=job_name,
+        docker_image=docker_image,
+        command=command,
+    )
```
What this is currently testing is simply that the calls via `boto3` (including the final submission) do not fail (which would have raised errors)

They do not guarantee the job itself succeeded, since that is separate from the submission

In a follow-up, I plan to leverage the helper function for table status updates to actually get these minimal tests to flag their completion at runtime, and then add an assertion (likely with some delay or repeat w/ timeout) that the status table was updated to 'completed' by the end of it all

h-mayorquin replied:
OK.

CodyCBakerPhD commented:
Evidence of AWS tests passing in CI: https://github.com/catalystneuro/neuroconv/actions/runs/10044953517/job/27761154548?pr=384

I'm disabling this hereafter from running on push to pull requests and limiting the matrix to reduce needless AWS costs

Otherwise, we only need to check it about once a week to ensure the `boto3` API doesn't deprecate structures we utilize here

h-mayorquin commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,283 @@ def estimate_s3_conversion_cost(
+
+def submit_aws_batch_job(
+    *,
+    job_name: str,
```
Can we make this optional and generate one automatically?

CodyCBakerPhD replied:
An internal `jobId` is automatically generated and is part of the info dict returned by this function; that is the internal one used by AWS operations

The `job_name` is a user/human readable tag related to the intent/usage/reference of that ID, so I think the user should specify it how they want

Otherwise the most we would do I think is utilize the `docker_image` as the base name and add a truncated UUID on the end to make it unique; but users would HAVE to grab the `info` dict returned to know how to find the name, which could be hard if they have lots of jobs running at once

h-mayorquin replied:
Explained on the meeting. Thanks.

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,283 @@ def estimate_s3_conversion_cost(
+    docker_image: str,
+    command: Optional[str] = None,
+    job_dependencies: Optional[List[Dict[str, str]]] = None,
+    region: str = "us-east-2",
```
It's more like it changes so much over time I wouldn't want to manually maintain a list

I could do something similar to what I do [here](https://github.com/catalystneuro/neuroconv/blob/main/src/neuroconv/tools/nwb_helpers/_configuration_models/_hdf5_dataset_io.py#L36) though where I can try to find some `boto3` API call to fetch currently available regions

h-mayorquin replied:
Probably not wroth it. I think this is fine and self-explanatory.

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,283 @@ def estimate_s3_conversion_cost(
+    command : str, optional
+        The command to run in the Docker container.
+        Current syntax only supports a single line; consecutive actions should be chained with the '&&' operator.
+    job_dependencies : list of dict
```
The only thing I'm cautious of here is offering too much information about the CURRENT `boto3` API given how much it can change even over the course of a year

My approach here is just to expose the ability to pass the arguments; perhaps we can simply link out to `boto3` documentation and tell users to reference that for exact syntax?

Most of the specifics for our own usage though will be handled in the following PR(s)

h-mayorquin replied:
Yeah, a link would be fine. Or just "refer for "x" on the boto3 documentation to see more" if we don't want to deal with possible dead links.

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,283 @@ def estimate_s3_conversion_cost(
+    compute_environment_name: str = "neuroconv_batch_environment",
+    job_queue_name: str = "neuroconv_batch_queue",
+    job_definition_name: Optional[str] = None,
+    minimum_worker_ram_in_gb: float = 4.0,
```
It's for our own API consistency, since all the iterator and backend configuration are in bytes

The reason for the RAM specification here is also in reference to iterative read/write usage through the iterators, so probably best that they both be specified in the same basis

We can go over this more in the meeting

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,283 @@ def estimate_s3_conversion_cost(
+
+
+def update_table_status(
+    *, submission_id: str, status: str, status_tracker_table_name: str = "neuroconv_batch_status_tracker"
```
```suggestion
    *, submission_id: str, status: str, status_tracker_table_name: str = "neuroconv_batch_status_tracker", region: str = "us-east-2",
```

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,283 @@ def estimate_s3_conversion_cost(
+        The random hash that was assigned on submission of this job to the status tracker table.
+    status : str
+        The new status value to update.
+    status_tracker_table_name : str, default: "neuroconv_batch_status_tracker"
+        The name of the DynamoDB table to use for tracking job status.
```
```suggestion
    status_tracker_table_name : str, default: "neuroconv_batch_status_tracker"
        The name of the DynamoDB table to use for tracking job status.
    region : str, default: "us-east-2"
        The AWS region to use for the job.
```

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,283 @@ def estimate_s3_conversion_cost(
+
+    dynamodb_resource = boto3.resource(
+        service_name="dynamodb",
+        region_name=region,
```
Good catch

This helper isn't actually tested or used until the follow-up, so I could have also removed from here altogether

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,285 @@ def estimate_s3_conversion_cost(
+            {"jobId": "job_id_1", "type": "N_TO_N"},
+            {"jobId": "job_id_2", "type": "SEQUENTIAL"},
+            ...
+        ]
```
```suggestion
        ]
        
        Refer to the boto3 API documentation for latest syntax:
        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/batch/client/submit_job.html
```

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,289 @@ def estimate_s3_conversion_cost(
+    compute_environment_name: str = "neuroconv_batch_environment",
+    job_queue_name: str = "neuroconv_batch_queue",
+    job_definition_name: Optional[str] = None,
+    minimum_worker_ram_in_gb: float = 4.0,
```
```suggestion
    minimum_worker_ram_in_gib: int = 4,
```

Maybe you're right though, since it does determine all kinds of allocation stuff in the cloud

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,289 @@ def estimate_s3_conversion_cost(
+    ]
+    resource_requirements = [
+        {
+            "value": str(int(minimum_worker_ram_in_gb * 1e3 / 1.024**2)),
```
```suggestion
            "value": str(int(minimum_worker_ram_in_gib * 1024)),
```

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,289 @@ def estimate_s3_conversion_cost(
+        The name of the job definition to use for the job.
+        Defaults to f"neuroconv_batch_{ name of docker image }",
+        but replaces any colons from tags in the docker image name with underscores.
+    minimum_worker_ram_in_gb : int, default: 4.0
+        The minimum amount of base worker memory required to run this job.
+        Determines the EC2 instance type selected by the automatic 'best fit' selector.
+        Recommended to be several GB to allow comfortable buffer space for data chunk iterators.
```
```suggestion
    minimum_worker_ram_in_gib : int, default: 4
        The minimum amount of base worker memory required to run this job.
        Determines the EC2 instance type selected by the automatic 'best fit' selector.
        Recommended to be several GiB to allow comfortable buffer space for data chunk iterators.
```

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,289 @@ def estimate_s3_conversion_cost(
+            computeEnvironmentName=compute_environment_name,
+            type="MANAGED",
+            state="ENABLED",
+            computeResources={
```
TODO: enable spot and expose % as keyword argument

CodyCBakerPhD replied:
default should be `100%` for minimum delay, but some users (such as us for big non-priority data churning) might want to manually set it lower to save costs

CodyCBakerPhD replied:
Actually after looking more into it, I recall that we don't intend to immediately support this because the processes can be easily interrupted by anyone else paying for on-demand, and these are such long processes usually that the risk of losing the instance (and thus any money spent to that point) is too high

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,289 @@ def estimate_s3_conversion_cost(
+    # Ensure job queue exists
+    current_job_queues = [queue["jobQueueName"] for queue in batch_client.describe_job_queues()["jobQueues"]]
+    if job_queue_name not in current_job_queues:
+        batch_client.create_job_queue(
```
TODO: expose argument for job state timeouts (when in 'runnable' not 'running' stalls out due to incompatibility of resource specification combined with spot %

CodyCBakerPhD replied:
Looking further into this, it seems all timeout conditions are for the event that the job exits `RUNNABLE` state and enters `RUNNING`. How long it stays in that state is entirely up to how much data we expect to process and how fast we expect it to be processed

So I'm OK with setting a somewhat large value here just for the sake of safety, but it doesn't solve the problem of jobs not deploying (or the problem of how to track that easily from our tools)

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,289 @@ def estimate_s3_conversion_cost(
+    # Update DynamoDB status tracking table
+    submission_id = submission_id or str(uuid4())
+    table_submission_info = dict(
+        id=submission_id, job_name=job_name, submitted_on=datetime.now().isoformat(), status="submitted"
```
TODO: include jobID from job submissio info for good measure

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/_aws.py:
```diff
@@ -30,3 +36,289 @@ def estimate_s3_conversion_cost(
+    job_dependencies = job_dependencies or []
+    container_overrides = dict(
+        # Set environment variables
+        environment=[
```
TODO: pass AWS credentials as well for status table updating

CodyCBakerPhD commented on src/neuroconv/tools/data_transfers/__init__.py:
```diff
@@ -1,14 +1,6 @@
-from ._aws import estimate_s3_conversion_cost
 from ._dandi import automatic_dandi_upload
 from ._globus import get_globus_dataset_content_sizes, transfer_globus_content
-from ._helpers import estimate_total_conversion_runtime
```
TBH these functions were probably never used by anyone so OK with removing them. Haven't even reviewed if they are still correct or not

CodyCBakerPhD commented:
I broke up some of the more repetitive bits blocks of code into private helpers and added docstrings to enhance long-term readability and maintainability

Also moved the new functions into a distinct submodule since this isn't only about 'data transfers' anymore

CodyCBakerPhD commented:
@h-mayorquin OK did a thorough deep dive with the new enhanced testing structure and fixed some issues from the running image, updated the subnets (these may have to be checked every few months-years) and documented some edge cases I discovered in the relevant docstrings

This should be ready for a second pass now

CodyCBakerPhD commented:
Evidence of enhanced tests passing: https://github.com/catalystneuro/neuroconv/actions/runs/10242963418/job/28333513602?pr=384

CodyCBakerPhD commented:
Breakdown of PR size:

+43 lines for a new distinct GitHub Action specific to these tests
+158 lines of tests, including thorough verification of final AWS job status
+640 lines of source code for the helper function and its private components; about 1/3 of this is docstrings recording the nuances of each particular step

h-mayorquin commented on src/neuroconv/tools/aws/_dynamodb.py:
```diff
@@ -0,0 +1,42 @@
+    """
+    import boto3
+
+    aws_access_key_id = os.environ["AWS_ACCESS_KEY_ID"]
```
missing the import os

h-mayorquin commented on src/neuroconv/tools/aws/_submit_aws_batch_job.py:
```diff
@@ -0,0 +1,640 @@
+    if len(lines_with_region) != 1:
+        return None
+
+    region = region_lines[0].split("=")[1].strip()
```
`lines_with_region` instead of `region_lines` here?

CodyCBakerPhD replied:
```suggestion
    region = lines_with_region[0].split("=")[1].strip()
```

h-mayorquin commented on src/neuroconv/tools/aws/_submit_aws_batch_job.py:
```diff
@@ -0,0 +1,640 @@
+            compute_environment_response[0]["status"] if len(compute_environment_response) == 1 else ""
+        )
+
+    if len(job_definition_response) != 1 and compute_environment_status != "VALID":
```
is this undefined?

CodyCBakerPhD commented on src/neuroconv/tools/aws/_dynamodb.py:
```diff
@@ -0,0 +1,42 @@
+"""Helper functions for operations on DynamoDB tables."""
+
```
```suggestion
import os
```

h-mayorquin commented on src/neuroconv/tools/aws/_submit_aws_batch_job.py:
```diff
@@ -0,0 +1,640 @@
+        dynamodb_client=dynamodb_client,
+        dynamodb_resource=dynamodb_resource,
+    )
+    role_info = _create_or_get_iam_role(iam_role_name=iam_role_name, iam_client=iam_client)
```
Let's cal this iam_role_info

CodyCBakerPhD replied:
Done

h-mayorquin commented on tests/test_minimal/test_tools/aws_tools.py:
```diff
@@ -0,0 +1,158 @@
+    )
+
+    # Wait for AWS to process the jobs
+    time.sleep(120)
```
Isn't this brittle? Can we query directly to see if the job is concluded?

CodyCBakerPhD replied:
Technically yes

If you want something better we could do the same retry mechanism as seen in source code

But this is such a simple job and has not had trouble so far just waiting a static amount of time

Whichever you prefer

h-mayorquin reviewed (commented):
Some more comments.

CodyCBakerPhD commented on src/neuroconv/tools/aws/_submit_aws_batch_job.py:
```diff
@@ -0,0 +1,640 @@
+        The name of the job definition to use for the job.
+    docker_image : str
+        The name of the Docker image to use for the job.
+    minimum_worker_ram_in_gib : int
```
Nope - there's a comment in the code about the serialization. It really should be an int always. I could relax and make it `MiB`, but I really think the conversion from `GiB` is a lot more convenient since you will always want RAM in that schedule

h-mayorquin commented on src/neuroconv/tools/aws/_submit_aws_batch_job.py:
```diff
@@ -0,0 +1,640 @@
+
+    Note that registering a job definition does not require either a compute environment or a job queue to exist.
+
+    Also note that registration is strict; once created, a job definition is tagged with ':< revision index >' and
```
TODO: For me.

h-mayorquin reviewed (approved):
TODO for me in another PR. Improve the string about the ARN changes.
